block autoplay shield-study extension

alastor0325 opened the pull request:
In order to understand how do people think about the "doorhanger" project which provides a prompt to ask autoplay permission for site, we need to write a shield-study extension to collect related data and response.

Commit:
update README.md

Commit:
config update

alastor0325 committed:
add content script for testing blocked autoplay media

alastor0325 committed:
add experiments autoplay API.

alastor0325 committed:
create a set for saving all URL user visited

alastor0325 committed:
modulize code to TabMonitor/TelemetrySender/ShieldStudyPing and main feature.

alastor0325 committed:
basic version sending telemetry ping

alastor0325 committed:
add autoplay check for tab.

alastor0325 committed:
minor changes

alastor0325 committed:
set preference

alastor0325 committed:
check autoplay permission status

alastor0325 committed:
add test button to send ping

alastor0325 committed:
add API event for notifying autoplay setting changes.

alastor0325 committed:
add observer for global setting changed.

alastor0325 committed:
Merge same type results into one ping

alastor0325 committed:
get blocked media count

alastor0325 committed:
move data probe to API side.

alastor0325 committed:
calculate the blocked autoplay count

alastor0325 committed:
send Telemetry ping back every 24 hours.

alastor0325 committed:
add type checking for telemetry scalar

alastor0325 committed:
remove useless api and function

alastor0325 committed:
user observer to detect autoplay

alastor0325 committed:
comment checkTabAutoplayStatus

alastor0325 committed:
move helper functions to public & adding salted-hash

alastor0325 committed:
add client id

alastor0325 committed:
correct the data type of 'allowAutoply'

alastor0325 committed:
remove content script

alastor0325 committed:
add different variations.

alastor0325 committed:
change preference by variation.

alastor0325 committed:
remove setting listener in close()

alastor0325 committed:
remove all autoplay changed code

alastor0325 committed:
change log

alastor0325 committed:
move initialization to congifure

alastor0325 committed:
remove hasAutoplayMediaContent().

alastor0325 committed:
add unit test

alastor0325 committed:
fix lint errors.

alastor0325 committed:
modify extension zip name

alastor0325 committed:
modify README

alastor0325 committed:
modify details of package.json

alastor0325 committed:
modify DEV.md

alastor0325 committed:
modify TELEMETRY.md

alastor0325 committed:
modify TELEMETRY.md

alastor0325 committed:
modify homepage

alastor0325 committed:
remove testing property.

alastor0325 committed:
add branch info in ping

alastor0325 committed:
TESTPLAN is TBD.

alastor0325 committed:
change autoplay prompt layout by different branches.

alastor0325 committed:
update TODO and change attribute name

alastor0325 committed:
remove api-child

alastor0325 committed:
some clean-up and minor changes

alastor0325 committed:
remove comment

alastor0325 committed:
remove test button.

alastor0325 committed:
update manifest description.

alastor0325 committed:
update description in schema.

alastor0325 committed:
remove comment

alastor0325 committed:
Update TESTPLAN.md

alastor0325 committed:
minor change in TESTPLAN.md

alastor0325 committed:
use sha256 as hashing function.

alastor0325 committed:
only register setting listener when the foreground tab is 'about:setting#privacy'.

alastor0325 committed:
fix lint error.

johannhof commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,337 @@
+
+    const count = accumulatedCount - this.blockedAudibleAutoplayCount;
+    if (count < 0) {
+      console.log("### Error : count should not be negative.");
```
What are you trying to do here? Just logging "Error" seems pretty inconsequential, either this is an issue (in which case you should throw an error or return 0) or it's not.

alastor0325 replied:
Thank you, I would return 0 here.

johannhof commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,337 @@
+  async constructPayload(payloadType) {
+    const payload = {
+      id: this.pingId++,
+      clientId: await getTelemetryId(),
```
Don't you already get the client id through the `addClientId: true` option in `submitExternalPing`?

alastor0325 replied:
Thank for reminding, I will remove `addClientId: true` because I want to store all the information in the payload part.

johannhof commented on src/feature.js:
```diff
@@ -1,144 +1,184 @@
+  return hashHex;
+}
+
+function isSupportURLProtocol(url) {
```
nit: `isSupportedURLProtocol`

alastor0325 replied:
fixed, thanks!

johannhof commented on src/feature.js:
```diff
@@ -1,144 +1,184 @@
+  return !!(url.match(/^(http(s?):\/\/)/im));
+}
+
+function getBaseDomain(url) {
```
I think we have a slightly different understand of "base" domain, e.g. if we're on `developer.mozilla.org`, for me the baseDomain would be `mozilla.org`, in this regex it's `developer.mozilla.org`.

That's probably what you want, but then you should consider using `new URL(url).hostname` (or `.host`)

https://developer.mozilla.org/en-US/docs/Web/API/HTMLHyperlinkElementUtils/hostname

alastor0325 replied:
Will change this function name to `getHostName()`, because what we save in the setting whitelist are host name, not base domain. Thank you!

alastor0325 replied:
After testing, `URL(url)` only works if the input url is complete url. However, when user change the whitelist website, the setting page would automatically change user's input to hostname and send it as input parameter to us. 

eg. enter `https://github.com/alastor0325/autoplay-shield-study/pull/3` in whitelist setting page, the result would be `https://github.com/` which is not acceptable input for function `URL()`.

Therefore, I can't use `URL(url).hostname` here.

johannhof commented on src/feature.js:
```diff
@@ -1,144 +1,184 @@
+async function getBaseDomainHash(url) {
+  const baseDomain = getBaseDomain(url);
+  if (baseDomain.length === 0) {
+    return hash;
```
This surely doesn't work with `const`, does it?

alastor0325 replied:
@johannhof why can not use `const` here? I won't change the `baseDomain` after getting it.

johannhof replied:
I was referring to returning `hash` before it is defined with const :)

I guess this code has changed enough that it doesn't matter anymore.

johannhof commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,337 @@
+    console.log("submitTelemetryPing to server");
+    console.log(data);
+    const telOptions = { addClientId: true, addEnvironment: true };
+    return TelemetryController.submitExternalPing("block-autoplay", data, telOptions);
```
Out of curiosity, why are you using `submitExternalPing` directly, instead of the Shield Telemetry?

alastor0325 replied:
Do you mean `browser.study.sendTelemetry()`? Because it only supports flat json format, but mine is nested json format.

johannhof replied:
Ah, okay. We usually just flatten our JSON. I think this is up to you but make sure that Shield people sign this off.

johannhof commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,337 @@
+            };
+
+            if (data === "added") {
+              autoplayState = (subject.capability === PERM_ACTION.ALLOW_ACTION) ?
```
nit: Replace this with `Ci.nsIPermissionManager.ALLOW_ACTION` and delete the `PERM_ACTION` enum above.

alastor0325 replied:
Thanks!

johannhof commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,337 @@
+          variation === "allow-and-remember") {
+        return [allowAction, denyAction];
+      }
+      return [denyAction, allowAction];
```
I just wanted to point out that, while this is fine for the experiment, we will not change the meaning of the default button of any doorhanger in m-c.

This is one of the most important aspects of consistency for users in these doorhangers. We must retain the intuitive understanding of users that pressing the blue button means "accept" and pressing the gray button means "deny".

I have commented this on the PHD.

alastor0325 replied:
Thank for suggestion, we will discuss this issue in our team meeting.

johannhof commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,337 @@
+        audibleAutoplayOccurred: new EventManager(context, "autoplay.audibleAutoplayOccurred", fire => {
+          const autoplayObs = (subject, topic, data) => {
+            fire.async(tabManager.getWrapper(subject).id,
+                       subject.linkedBrowser.currentURI.spec.toString());
```
nit: isn't `spec` already a string? Or do you want to explicitly copy it?

alastor0325 replied:
You're right, I will remove `toString()`. Thank you!

johannhof commented on src/feature.js:
```diff
@@ -1,144 +1,184 @@
+    return hash;
+  }
+
+  const hash = await sha256(baseDomain + gId);
```
I'd just like to point out for posterity that this is not exactly what the PHD says:

> Assign etld a local id, for this etld, when the experiment determines a page has videos to be blocked.

This means instead of a unique hash, it wants you to persist an ID for each domain, which is even more privacy-preserving for the user because there's no way for us to brute-force that.

We're having the same discussion for FastBlock right now, though generally I think it makes a lot of sense and could be adopted as a standard mechanism for Shield studies: https://github.com/mozilla/FastBlockShield/issues/43

If your team is okay with your approach then that works for me.

alastor0325 replied:
OK, I will change my implementation. Thanks!

johannhof commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,337 @@
+
+        setPreferences: (variation) => {
+          this.branch = variation;
+          Preferences.set({
```
Are you ever restoring these Preferences to default? I can't see where by code inspection and in my testing they weren't restored either...

alastor0325 replied:
Do you mean restore the prefence to default when the study ends? I will add it in `cleanup()`. Thanks!

alastor0325 replied:
@johannhof here I have some questions.

In the shield-study framework, does the `variation` is only selected by framework when first time the shield-study run? does the value of variation is from one of `weightedVariations` which I define in the `baseStudySetup`?

I'm testing to permanantly install my extension and then uninstall it on Nightly, and I found that every time I install my extension, I would get the same `variation`. I expect that I should get different `variation` everytime when I reinstall it.

Is that a normal situation? Thanks!

johannhof replied:
Good question for @gregglind...

johannhof reviewed (changes requested):
Generally this looks solid, but there are a couple of issues that we should fix before this can ship, see below for more details :)

johannhof commented on src/feature.js:
```diff
@@ -1,144 +1,184 @@
+    this.tabsMonitor = new TabsMonitor();
+    this.tabsMonitor.configure(this);
+    const sendPingIntervalMS = 1 * 24 * 60 * 60 * 1000;
+    this.sendPingsScheduler(sendPingIntervalMS);
```
What I also wanted to remark: There are a bunch of scenarios in which users would not send pings when just scheduling them via setInterval. This could skew your results in favor of users who rarely restart their browser. Sending the telemetry on experiment end helps a bit but you should still consider storing the timeout value (or using shield telemetry which AFAIK aggregates telemetry probes before sending).

I won't require it, however, just wanted to note this.

alastor0325 replied:
@johannhof Sorry, I don't get it. Why users who rarely restart their browser would affect the the ping sending by `setInterval()`? 

Here, I want to send the ping when one of following conditions happens (1) close browser (2) not close browser, but time passes a day. Therefore, I've add the `sendTelemetryPings()` in the `AutoplayAPI::onShutdown()`.

Thanks!

johannhof replied:
Ok, yeah, sounds like I was wrong about this :)

alastor0325 committed:
Review modification

johannhof commented on src/feature.js:
```diff
@@ -1,147 +1,186 @@
-  configure(studyInfo) {
-    const feature = this;
-    const { variation, isFirstRun } = studyInfo;
+var hostNamesMap = new Map();
```
Just a note that while the PHD does not explicitly mention this, you probably want to persist these IDs in WebExtension storage to make it survive through restarts. I asked for clarification in the PHD.

alastor0325 replied:
OK, will do, thank you!

johannhof commented on src/feature.js:
```diff
@@ -1,147 +1,186 @@
+  if (!hostNamesMap.has(hostName)) {
+    hostNamesMap.set(hostName, generateRandomString());
+  }
+  const hash = await sha256(hostName + hostNamesMap.get(hostName));
```
There is no need to hash or salt anything, just submit the local ID, since the hostname can not be derived from the ID without accessing the local map.

alastor0325 replied:
OK, will do, thank you!

johannhof commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,332 @@
+  return p;
+}
+
+function errorHandler(err) {
```
It would probably be more idiomatic to use `Cu.reportError` instead of this. https://searchfox.org/mozilla-central/rev/83562422ecb0f5683da7a9a26ce05ce62bc0c882/js/xpconnect/idl/xpccomponents.idl#157

alastor0325 replied:
OK, thank you!

johannhof commented on src/feature.js:
```diff
@@ -1,147 +1,186 @@
+  async cleanup() {
+    this.tabsMonitor.clear();
+    // restoring to the default option.
+    browser.autoplay.setPreferences("allow-and-remember");
```
Hmm, this is assuming that `browser.autoplay.setPreferences("allow-and-remember");` will always reset to the default prefs. That assumption could become false if your team decides to uplift a patch that changes the default pref.

To be certain, I would recommend going the extra mile and adding a `browser.autoplay.clearPreferences` function.

alastor0325 replied:
I am not sure whether we would change the default layout of doorhanger in the future, but you comment reminds me that I should also remember the current value of preference and restore them when the study ends. Thank you!

johannhof reviewed (approved):
This looks fine to me, the remaining issues are mostly things to figure out with your team and I don't think I would absolutely need to have another look at this.

In doubt feel free to ask for another quick review/feedback, though :)

Thanks!

alastor0325 committed:
Save host name in local storage, remove hashing, and restore user preference in the end of study

alastor0325 committed:
modify JSON format

Based on the discussion [1], change our json format.

[1] https://github.com/mozilla-services/mozilla-pipeline-schemas/pull/187#issuecomment-414830592

alastor0325 committed:
Only send 'counter' ping for control branch

the control testing branch do not need to collect prompt and setting changes data.

alastor0325 committed:
remove unnecessary log.

johannhof commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,365 @@
+              setting[key] = item[key];
+            }
+          }
+          Preferences.set(setting);
```
You can just use `Preferences.reset` instead, no?

https://searchfox.org/mozilla-central/rev/3fa761ade83ed0b8ab463acb057c2cf0b104689e/toolkit/modules/Preferences.jsm#199

alastor0325 replied:
@johannhof But I think `Preferences.reset` would clear all the user-set value to default, instead of changing value to the value which users have before starting shield study, right?

alastor0325 committed:
move clientId to the part of common ping format

According to [1], the clientId should be a part of outer common ping format, instead
of being a part of payload.

[1] https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/data/common-ping.html#common-ping-format

alastor0325 committed:
Update preparations config

Should also change the config of "xpinstall.signatures.required" and "extensions.legacy.enabled" when the extension hasn't be signed yet.

alastor0325 committed:
Remove enviroment in common ping.

alastor0325 committed:
Adding new ping attribute "interact"

User can have different ways to interact with the prompt, they can either click a button, hit the
‘escape’ key, or ignore it and do nothing.

Therefore, we add new attribute `interact` with string enum value {interact, escape, ignore}.

In addition, `rememberCheckbox` and `allowAutoPlay` will only be included if the value of `interact` is `interact`.

alastor0325 committed:
Update telemetry.md

alastor0325 committed:
Add new attribute `responseTime` in `promptChanged`.

`responseTime` is measuring how long the prompt shows before users make their response.

alastor0325 committed:
Update Telemetry.md for adding `responseTime`

`responseTime` is measuring how long the prompt shows before users make their response.

alastor0325 committed:
Change testing cohorts

According to our final decision, we would change the number of testing cohorts to 4.

1. control (always allow)
2. block (always block)
3. allow-and-remember
4. allow-and-notRemember

alastor0325 committed:
update dependencies

alastor0325 committed:
add ending survey

alastor0325 committed:
Make extension hidden in about:addons

alastor0325 committed:
handle uninstallation by Normandy

alastor0325 committed:
Update TESTPLAN.md

Modify the description about testing branches.

alastor0325 committed:
Add learnForMore link.

alastor0325 committed:
clean-up all stuffs before uninstall extension.

rhelmer commented on src/feature.js:
```diff
@@ -1,147 +1,191 @@
-  configure(studyInfo) {
-    const feature = this;
-    const { variation, isFirstRun } = studyInfo;
+function generateRandomString() {
```
There's a service in Firefox privileged JS to generate UUIDs if you need that.

alastor0325 replied:
I thought that we can only use privileged JS code in experiments API, not in background script?

rhelmer commented on src/feature.js:
```diff
@@ -1,147 +1,191 @@
+}
+
+function getHostName(url) {
+  return url.match(/^(?:https?:\/\/)?(?:[^@\n]+@)?(?:www\.)?([^:\/\n]+)/im)[1];
```
Could use the standard [URL API](https://developer.mozilla.org/en-US/docs/Web/API/URL) to tell if a URL is valid, extract the hostname etc. instead of having to do your own regex.

alastor0325 replied:
In `about:preferences`, user can set any url into "Autoplay - Exception" which might include some formats which URL API doesn't accept. (eg. "123.com" is not a valid input format for URL API.)

Therefore, it means I need to do an additional transformation and check for every input, it would be more complicated. I would prefer to use my own regex rather than doing that.

rhelmer commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,397 @@
+  return Services.strings.createBundle("chrome://browser/locale/browser.properties");
+});
+
+function _once(target, name) {
```
Hm. I might be misunderstanding the intent here, but [`addEventListener` takes a boolean named `once`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) if you just want an event listener to fire a single time:

alastor0325 replied:
Since I want to use await pattern [here](https://github.com/alastor0325/autoplay-shield-study/blob/995dd3286df3ae659985a899936ceda691b47849/src/privileged/autoplay/api.js#L370) for waiting until the event happen, `_once` is just a wrapper helper function.

rhelmer commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,397 @@
+const { AddonStudies } = ChromeUtils.import("resource://normandy/lib/AddonStudies.jsm", {});
+const { EventManager } = ExtensionCommon;
+
+XPCOMUtils.defineLazyGetter(this, "gBrowserBundle", function() {
```
You don't need to import `XPCOMUtils` for this anymore, `ChromeUtils` supports it like this:
```
ChromeUtils.defineModuleGetter(this, "MyModule",
                               "resource://MyModule.jsm");
```

alastor0325 replied:
But here I don't define a module, I defined a getter for the variable. It seems not equal to the usage of `defineModuleGetter()`?

rhelmer commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,397 @@
+ChromeUtils.import("resource:///modules/SitePermissions.jsm");
+ChromeUtils.import("resource://gre/modules/ExtensionCommon.jsm");
+ChromeUtils.import("resource://gre/modules/ExtensionStorage.jsm");
+ChromeUtils.import("resource://gre/modules/PopupNotifications.jsm");
```
I suspect you could make more of these lazy - basically anything that has the possibility of not being used is good to not directly import.

alastor0325 replied:
Ok, thank you!

rhelmer commented on src/privileged/autoplay/api.js:
```diff
@@ -0,0 +1,397 @@
+function getTelemetryId() {
+  const id = TelemetryController.clientID;
+  if (id === undefined) {
+    return CID.ClientIDImpl._doLoadClientID();
```
Why not just `await ClientID.getClientID()`? It's already [doing this for you](https://searchfox.org/mozilla-central/rev/881a3c5664ede5e08ee986d76433bc5c4b5680e6/toolkit/modules/ClientID.jsm#173) and better to not call private methods in other modules :)

alastor0325 replied:
OK! Thank you!

alastor0325 replied:
Ah, I found that this function should be removed, since I have already made TelemetryController to generate `clientId`.

rhelmer reviewed (approved):
lgtm overall - I have not tried running it locally etc. just looked over the code. I don't think any of these suggestions are blocking but could probably make things clearer here and there.

alastor0325 committed:
update doorhanger's check-box in private browsing mode.

alastor0325 committed:
only set preferences when first time extension runs.

alastor0325 committed:
remove 'getTelemetryId()'.

In [1], we have already made TelemetryController to generate ID, we don't need this function anymore.

[1] https://github.com/alastor0325/autoplay-shield-study/commit/e45a4c6217a2adbe6680901d5ac6dbeb4510f659

alastor0325 committed:
use ChromeUtils.defineModuleGetter() to lazily load modules.

alastor0325 committed:
clear tabMonitor if it exists.

alastor0325 committed:
Set doorhanger's checkbox properly after user restarts browser.

alastor0325 committed:
Update version to 2.0.1

update version number in order to get second signed version extention.

alastor0325 committed:
update TELEMETRY.md.

alastor0325 committed:
remove testing flag

alastor0325 committed:
change version number

alastor0325 committed:
Modify clearPreferences()'s schema.
